catalog/transformer/catalog-transformer-streaming-lib/src/main/java/org/codice/ddf/transformer/xml/streaming/lib/MetacardTypeRegister.java

```diff
@@ -0,0 +1,163 @@
+    }
+
+    public MetacardType getMetacardType() {
+        if (metacardType == null) {
+            setMetacardType();
+        }
+
+        return metacardType;
+    }
+
+    private synchronized void registerMetacardType(MetacardType metacardType) {
+        unregisterMetacardType(metacardType);
+
+        Dictionary serviceProperties = new Hashtable();
+        serviceProperties.put("name", metacardType.getName());
+
+        ServiceRegistration serviceRegistration =
+                getContext().registerService(MetacardType.class.getName(),
+                        metacardType,
+                        serviceProperties);
+
+        this.metacardTypeServiceRegistrations.put(metacardType.getName(), serviceRegistration);
+    }
+
+    private synchronized void unregisterMetacardType(MetacardType metacardType) {
+        if (metacardType == null) {
+            return;
+        }
+
+        ServiceRegistration serviceRegistration =
+                metacardTypeServiceRegistrations.get(metacardType.getName());
+        if (serviceRegistration != null) {
+            serviceRegistration.unregister();
+            metacardTypeServiceRegistrations.remove(metacardType.getName());
+        }
+    }
+
+    /**
+     * Setter to set the list of all {@link SaxEventHandlerFactory}s. Usually called from a blueprint,
+     * and the blueprint will keep the factories updated
+     *
+     * @param saxEventHandlerFactories a list of all SaxEventHandlerFactories (usually a list of all
+     *                                 factories that are available as services)
+     */
+    public void setSaxEventHandlerFactories(List<SaxEventHandlerFactory> saxEventHandlerFactories) {
+        this.saxEventHandlerFactories = saxEventHandlerFactories;
+    }
+
+    /**
+     * Setter to set the configuration of SaxEventHandlers used to parse metacards
+     *
+     * @param saxEventHandlerConfiguration a list of SaxEventHandlerFactory ids
+     */
+    public void setSaxEventHandlerConfiguration(List<String> saxEventHandlerConfiguration) {
+        this.saxEventHandlerConfiguration = saxEventHandlerConfiguration;
+    }
+
+    public void setId(String id) {
+        this.id = id;
+    }
+}
+
+class DynamicMetacardType implements MetacardType {
+
+    Set<AttributeDescriptor> attributeDescriptors;
+
+    String name;
+
+    public DynamicMetacardType(Set<AttributeDescriptor> attDesc, String name) {
+        this.attributeDescriptors = attDesc;
+        this.name = name + ".metacard";
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    @Override
+    public Set<AttributeDescriptor> getAttributeDescriptors() {
+        return attributeDescriptors;
+    }
+
+    @Override
+    public AttributeDescriptor getAttributeDescriptor(String attributeName) {
+        return attributeDescriptors.stream()
+                .filter(p -> p.getName()
+                        .equals(attributeName))
+                .collect(Collectors.toList())
```

Comment:
If you just want the first, don't waste the time processing the whole collection and building a new List. Just do

```java
return attributeDescriptors.stream()
    .filter(p -> p.getName().equals(attributeName))
    .findFirst()
    .orElse(null);
```
That should do it.